uat/testing-features/src/main/java/com/aws/greengrass/aws/S3Bucket.java

```diff
@@ -0,0 +1,57 @@
+package com.aws.greengrass.aws;
+
+import com.amazonaws.services.s3.model.Bucket;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.aws.greengrass.testing.resources.AWSResource;
+import lombok.Data;
+import lombok.extern.log4j.Log4j2;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+import static com.aws.greengrass.aws.IotShadowResources.randomName;
+
+@Data
+@Log4j2
+public class S3Bucket implements AWSResource<AmazonS3> {
+    boolean deleted = false;
+    private String bucketName;
+
+    public S3Bucket(String bucketName) {
+        this.bucketName = bucketName;
+    }
+
+    public static S3Bucket create(AmazonS3 amazonS3, S3BucketSpec spec) {
+        if (spec.bucketName == null) {
+            spec.setBucketName(randomName());
+        }
+        log.atDebug().log("Creating a S3 Bucket {}", spec.bucketName);
+        Bucket b = amazonS3.createBucket(spec.bucketName);
+        return new S3Bucket(spec.bucketName);
+    }
+
+    @Override
+    public void remove(AmazonS3 amazons3) {
+        if (deleted) {
+            return;
+        }
+        log.atDebug().log("Removing S3 Bucket {}", bucketName);
+        // Assuming that S3 bucket is not versioned in this case
+        // removing objects from bucket
+        ObjectListing object_listing = amazons3.listObjects(bucketName);
+        while (true) {
+            for (S3ObjectSummary summary : object_listing.getObjectSummaries()) {
+                amazons3.deleteObject(bucketName, summary.getKey());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code is using a single API operation, but a batch API operation exists. You could help improve efficiency by using the batch operation instead. The batch operation is deleteObjects.